Implement work_group_static / work_group_scratch_memory (#15061)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_work_group_scratch_memory.asciidoc

= sycl_ext_oneapi_work_group_scratch_memory

:source-highlighter: coderay
:coderay-linenums-mode: table

// This section needs to be after the document title.
:doctype: book
:toc2:
:toc: left
:encoding: utf-8
:lang: en
:dpcpp: pass:[DPC++]

// Set the default source code type in this document to C++,
// for syntax highlighting purposes.  This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

== Notice

[%hardbreaks]
Copyright (C) 2024 Intel Corporation.  All rights reserved.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

> This extension is written against the SYCL 2020 revision 8 specification.  All

Reviewer: ```suggestion
This extension is written against the SYCL 2020 revision 9 specification.  All
```

Let's use the most recent.  Same for the other spec in this PR.

Author: I applied suggestion

references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

The following extensions are required:

- link:../experimental/sycl_ext_oneapi_properties.asciidoc[sycl_ext_oneapi_properties]

- link:../experimental/sycl_ext_oneapi_kernel_properties.asciidoc[sycl_ext_oneapi_kernel_properties]

== Status

This is an experimental extension specification, intended to gather community
feedback.  Interfaces defined in this specification may not be implemented yet
or may be in a preliminary state.  The specification itself may also change in
incompatible ways before it is finalized.  *Shipping software products should

> not rely on APIs defined in this specification.*

Reviewer: This is not the right text for an experimental extension.  In particular, the part about "Interfaces defined in this specification may not be implemented yet or may be in a preliminary state" should not be the case.  We expect all interfaces of an experimental extension to be implemented as documented.  See here for the correct text:

https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc#status

Author: fixed, thanks

== Overview

This extension adds a way to allocate device local memory, without explicitly passing a
kernel argument: `get_work_group_scratch_memory`. It provides access to a dynamically sized
buffer without passing it as an argument to the kernel.
Device local memory is memory that is shared by all work-items in a work-group.
The behavior is similar to the usage of unbounded array with the CUDA `+__shared__+` keyword.

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the

> macro `SYCL_EXT_ONEAPI_WORK_GROUP_DYNAMIC` to one of the values defined in the

Reviewer: ```suggestion
macro `SYCL_EXT_ONEAPI_WORK_GROUP_SCRATCH_MEMORY` to one of the values defined in the
```

The macro name should match the name of the extension.

Author: oh sorry, I forgot to update this when I renamed the ext, thanks

table below.  Applications can test for the existence of this macro to
determine if the implementation supports this feature, or applications can test
the macro's value to determine which of the extension's features the
implementation supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

=== `get_work_group_scratch_memory` function

The `get_work_group_scratch_memory` function provides access
to a dynamically allocated buffer in the device local memory.

[source,c++]
----

> void* get_work_group_scratch_memory()

Reviewer: The synopsis should show the namespace.

Author: fixed, thanks

----

> _Constraints_: `T` must be trivially constructible and trivially destructible.

Reviewer: Is this left over from a previous iteration?  There is no template type `T`.

Author: it is, fixed, thanks

the run-time size of the device local memory allocation associated with
all `get_work_group_scratch_memory` calls.

[source,c++]
----
namespace sycl::ext::oneapi::experimental {

struct work_group_scratch_size {
  constexpr work_group_scratch_size(size_t bytes) : value(bytes) {}
  size_t value;
}; // work_group_scratch_size

using work_group_scratch_size_key = work_group_scratch_size;

template <> struct is_property_key<work_group_scratch_size_key> : std::true_type {};

} // namespace sycl::ext::oneapi::experimental
----

|===
|Property|Description

|`work_group_scratch_size`
|The `work_group_scratch_size` property describes the amount of dynamic
device local memory required by the kernel in bytes.

|===

=== Total allocation check

If the total amount of device local memory requested (i.e., the sum of
all memory requested by `local_accessor`, `group_local_memory`,
`group_local_memory_for_overwrite`, `work_group_static` and `work_group_scratch_size`) exceeds a device's
local memory capacity (as reported by `local_mem_size`) then the implementation
must throw a synchronous `exception` with the `errc::memory_allocation` error

> code from the kernel invocation command (e.g. `parallel_for`).

Reviewer: This list of ways to allocate local memory is incomplete.  I'd suggest rewording to something like this:

> If the total amount of device local memory requested exceeds a device’s local memory capacity as reported by `info::device::local_mem_size` then the implementation must throw a synchronous `exception` with the `errc::memory_allocation` error code from the kernel invocation command (e.g. `parallel_for`).  This check must take all APIs that allocation device local memory into account, whether via the `work_group_scratch_size` property or other APIs such as `local_accessor`.

Author: I applied your suggestion to both extensions

> ==== Usage examples

Reviewer: ```suggestion
== Example
```

The "Example" section should be a peer of "Specification" because the examples are not a normative part of the specification.

Author: indeed, fixed both extensions

> ===== Allocations with size unknown at compile-time

Reviewer: Delete this heading.  All uses of this extension allocate memory with a size that is unknown at compile time.

Author: I fixed both extensions

[source,c++]
----

> using namespace syclex = sycl::ext::oneapi::experimental;

Reviewer: ```suggestion
namespace syclex = sycl::ext::oneapi::experimental;
```

Author: I applied suggestion

...

q.parallel_for(sycl::nd_range<1>{N, M},
  syclex::properties{syclex::work_group_scratch_size(M * sizeof(int))},
  [=](sycl::nd_item<1> it) {
 auto ptr = get_work_group_scratch_memory();

>  auto ptr2 = get_work_group_scratch_memory();

Reviewer: Add a namespace qualifier to these calls.

Author: fixed, thanks
